File: 530.minimum-absolute-difference-in-bst.py
#
# @lc app=leetcode id=530 lang=python3
#
# [530] Minimum Absolute Difference in BST
#


# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        self.ans = float("inf")
        self.traverse(root, float("inf"), float("-inf"))
        return self.ans

    def traverse(self, node, upperbound, lowerbound):
        if node:
            self.ans = min(self.ans, upperbound - node.val, node.val - lowerbound)
            self.traverse(node.left, node.val, lowerbound)
            self.traverse(node.right, upperbound, node.val)


# @lc code=end

File: 530.minimum-absolute-difference-in-bst.go
/*
 * @lc app=leetcode id=530 lang=golang
 *
 * [530] Minimum Absolute Difference in BST
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

package main

import "math"

func getMinimumDifference(root *TreeNode) int {
	nodeVals := inorder(root)
	minDiff := math.MaxInt
	n := len(nodeVals)
	for i := 0; i < n-1; i++ {
		diff := nodeVals[i+1] - nodeVals[i]
		if diff < minDiff {
			minDiff = diff
		}
	}
	return minDiff
}

func inorder(root *TreeNode) []int {
	if root == nil {
		return []int{}
	}
	return append(append(inorder(root.Left), root.Val), inorder(root.Right)...)
}

// @lc code=end
